### ISSUE TITLE
[Bug]: axes.set_xlim with string dates raises when plotting with datetimes

### ISSUE THREAD

**Issue by mroeschke** [matplotlib/matplotlib#25219]
### Bug summary

pandas is seeing recent failures with the 3.7.0 release when using `set_xlim` with strings representing dates and plotting on the x axis with dates

### Code for reproduction

```python
In [1]: import matplotlib.pyplot as plt

In [2]: dates = np.array(['2000-01-01T00:00:00.000000000', '2000-01-02T00:00:00.000000000'], dtype='datetime64[ns]')

In [3]: values = np.arange(2)

In [4]: _, ax = plt.subplots()

In [5]: ax.plot(dates, values)
Out[5]: [<matplotlib.lines.Line2D at 0x7fba780ae130>]

In [6]: ax.set_xlim("1/1/1999", "1/1/2001")
Traceback (most recent call last):
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/axis.py", line 1736, in convert_units
    ret = self.converter.convert(x, self.units, self)
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/category.py", line 57, in convert
    unit.update(values)
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/category.py", line 216, in update
    _api.check_isinstance((str, bytes), value=val)
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/_api/__init__.py", line 93, in check_isinstance
    raise TypeError(
TypeError: 'value' must be an instance of str or bytes, not a int

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/cbook/__init__.py", line 304, in process
    func(*args, **kwargs)
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/axes/_base.py", line 2455, in _unit_change_handler
    line.recache_always()
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/lines.py", line 652, in recache_always
    self.recache(always=True)
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/lines.py", line 656, in recache
    xconv = self.convert_xunits(self._xorig)
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/artist.py", line 281, in convert_xunits
    return ax.xaxis.convert_units(x)
  File "/Users/matthewroeschke/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/matplotlib/axis.py", line 1738, in convert_units
    raise munits.ConversionError('Failed to convert value(s) to axis '
matplotlib.units.ConversionError: Failed to convert value(s) to axis units: array(['2000-01-01T00:00:00.000000000', '2000-01-02T00:00:00.000000000'],
      dtype='datetime64[ns]')
Out[6]: (0.0, 1.0)

In [7]: import matplotlib; matplotlib.__version__
Out[7]: '3.7.0'
```

### Actual outcome

(The iPython example above shows the outcome)

### Expected outcome

No exception traceback.

### Additional information

This appeared to work in the 3.6.x series. I could no find documentation regarding this functionality not being supported

### Matplotlib Version

3.7.0

### Python version

3.8

### Installation

conda

**Comment by jklymak** [comment]
I do not think that we ever converted dates in strings like that.  Sorry to bounce this back but 99.9% sure this is a pandas issue.  

```
import matplotlib.pyplot as plt
import numpy as np

dates = np.array(['2000-01-01T00:00:00.000000000', '2000-01-02T00:00:00.000000000'], dtype='datetime64[ns]')

values = np.arange(2)

_, ax = plt.subplots()

ax.plot(dates, values)
ax.set_xlim("1/1/1999", "1/1/2001")

plt.show()
```

fails the same way on 3.6.2.

**Comment by jklymak** [comment]
BTW, we do handle datetime64 and datetime so 

```python
ax.set_xlim(np.datetime64("1999-01-01"), np.datetime64("2001-01-01"))
```

works fine

**Comment by ksunden** [comment]
I can confirm that while I think the test as written here is invalid (we don't expect strings to parse as dates by default), pandas tests do fail with mpl 3.7 and it bisects to 31dee3ac15d347d552291d0a992d663b932a1dc7 from #23417, which was addressing #23416

The line in the tests is indeed setting xlim using strings, but that is a feature of Pandas' Converter, not the built in datetime/numpy support for matplotlib.

In the datetimelike test, the axes has a `pandas.plotting._matplotlib.converter.PeriodConverter`, but `units` of `None`.

Setting the units (e.g. `ax.xaxis.set_units("UTC")`) in the test prior to `set_xlim` does in fact pass the test (though I'm not sure if that should be the "accepted" fix going forward).

The `test_series` test is the same except a `pandas.plotting._matplotlib.converter.DatetimeConverter`. 

The change from #23417 caused a tweak from "one of converter or units is not None" to "both of converter and units are not None" for calling `update_units` in `_process_unit_info`.

**Comment by jklymak** [comment]
I'm not following how the axis got a converter without getting units - they both get set in `axis.update_units`.  Is there a self- contained example (even if it uses pandas)?

**Comment by ksunden** [comment]
Specifically, update_units is not being called in the affected method because of that change

**Comment by jklymak** [comment]
Sure but I don't understand how the axis got in that state.

**Comment by ksunden** [comment]
```python
from pandas import Series, date_range
import matplotlib.pyplot as plt
import numpy as np

rng = date_range("1/1/2000", "3/1/2000")
ser = Series(np.random.randn(len(rng)), rng)
_, ax = plt.subplots()
ax = ser.plot(ax=ax)
print(type(ax.xaxis.converter))  # pandas PeriodConverter
print(ax.xaxis.units)  # None
ax.set_xlim("1/1/1999", "1/1/2001")
```

Here is a pretty minimal example (mostly just the failing test, but with some test fixtures that are not needed to reproduce removed)

Interestingly, after `ax.set_xlim`, the xaxis converter is a StrCategoryConverter instead of the pandas converter it had been.

Part of this may be the `default_units`. Pandas does not override DateConverter's default_units:

https://github.com/matplotlib/matplotlib/blob/39070aeaab732eca1b5b7f7b478fa8b72dab8507/lib/matplotlib/dates.py#L1864-L1882

```python
>>> rng[0].tzinfo  # is None by default
```

So even if `default_units` is called, units will be None, and the call to `update_units` in `_process_unit_info` will not be called.

So, in summary:

- Date converters (including Panda's subclass) do allow `None` as units
- The change to make scatter plots consistent with handling of units means that if units are None, behavior changes
- I cannot seem to reproduce the behavior with roughly equivalent non-pandas code:

```python
import matplotlib.pyplot as plt
import numpy as np

rng = np.arange(np.datetime64("2000-01-01"), np.datetime64("2001-01-01"), 12)
y = np.random.randn(len(rng))
_, ax = plt.subplots()
ax.plot(rng, y)
print(type(ax.xaxis.converter))  # SwitchableDateConverter
print(ax.xaxis.units)  # None
ax.set_xlim(np.datetime64("1999-01-01"), np.datetime64("2001-01-01"))
```


So I'm not sure what the next steps are, but I think I understand how it came to have a converter and `None` units, at the very least.

**Comment by jklymak** [comment]
Right - so I don't understand why pandas fails and we do not.  I guess I also don't know why we check for units are None now.  The fundamental issue is that None meant "not set" _and_ "don't care". If we _really_ need to check if units are set, then "don't care" should probably be an empty string.  For dates it could be "UTC" as that is default for datetime64 and other datetime with no tzinfo.  

But most important we should document all this and standardize it. Ideally it should all be rearchitected 😉

**Comment by ksunden** [comment]
A bit more info: The converter is reset to `StrCategoryConverter` _before_ the exception.

I believe that what is happening is that the call to `_process_unit_info` in `axes._set_lim` is overriding the unit info with mpl's default handling of strings (as categoricals).

https://github.com/matplotlib/matplotlib/blob/39070aeaab732eca1b5b7f7b478fa8b72dab8507/lib/matplotlib/axis.py#L1190

**Comment by jklymak** [comment]
OK, that may trace to something else where people felt that setting the limits on an empty chart should go through the units machinery.  So somehow that is not checking that there is already a converter and that it accepts strings...

**Comment by mroeschke** [comment]
Sorry for not providing a sufficient initial example (we don't have many knowledgeable maintainers anymore related to the matplotlib code).

> Part of this may be the default_units. Pandas does not override DateConverter's default_units:

If I am understanding what this is used for, some of the dates that can be plotted do not have timezone information so returning `None` here would still be appropriate or would a fix be to set a "placeholder" tzinfo like UTC?

**Comment by ksunden** [comment]
```diff
diff --git a/lib/matplotlib/dates.py b/lib/matplotlib/dates.py
index 4b7e5a99a7..cbcef16258 100644
--- a/lib/matplotlib/dates.py
+++ b/lib/matplotlib/dates.py
@@ -1876,10 +1876,10 @@ class DateConverter(units.ConversionInterface):
             pass
 
         try:
-            return x.tzinfo
+            return _get_tzinfo(x.tzinfo)
         except AttributeError:
             pass
-        return None
+        return _get_tzinfo(None)
 ```
 
This diff resolves the unit-related problems with pandas tests (the test related to the merging of Axes and Subplot is, of course, unaffected by this)

Essentially, this is greedily resolving the default tz unit (which is `rcParams.timezone`, which defaults to UTC. 

No mpl tests fail (though I probably wish to _add_ a test here)

**Comment by mroeschke** [comment]
Thanks so much for the investigation @jklymak and @ksunden

**Comment by ksunden** [comment]
Okay... I'm not opposed to adding the above patch which solves the narrow issue at play here with pandas specifically, but I'm also not sure I wish to say "default units cannot be None" in the general case.

The other oddity that is very relevant here is indeed the fact that Pandas' converters take strings, so the lookup of which converter from `munits.registry.get_converter` when handed string data does not get the same converter. I'm also not sure I _like_ that fact, but not sure I'm willing to say that it should be disallowed (e.g. even if we didn't use strings but used np.datetime64, that would also be true that the actual converter is different than the converter you get for setting xlim, but those should probably be allowed for downstream DateConverters, I think)

- #23417 was made because if you set the units before plotting, you still need to set the converter when a plotting artist is added.
    - `update_units` is called in a `_process_unit_info` which is called in several places, notably many of the artist functions of `Axes` and also important for this dicussion in `set_[xy]lim`.
    - `update_units` is also called separately when processing the `*args` for `Axes.plot`, which caused that to have special behavior that was prioritized for the PR.
    - When users set units _before_ plotting, the axis still has no converter, even though it has units, but you still need to call `update_units` to attach a converter when adding an artist.
    - so the logic was expanded from `~(units not None || converter not None)` to `~(units not None && converter not None)` which meant that when units were set we still call `update_units`, get a converter and everybody is happy (for this workflow)

Now for pandas, that breaks down a bit:

- The date converters have a default of None (should probably be allowed)
- Since the unit is None, we call `update_units`
- When handed strings in xlim, we say "Oh, you want a string categorical converter"
- since the converter it gets is different than what it had, it trusts the new info and resets the converter
- It then breaks when it tries to convert existing data


I tried simplifying the logic from the two conditionals above to simply `converter is None`, but that breaks the new test from #23417, and I'm not _fully_ sure why, and not sure it is still true that it is semantically what we want.

So I ask the following questions:

What, precisely, do we mean when we say the units are `None`? 
  - Does it mean we should disregard any unit-adjacent things? (probably not)
  - Does it mean the converter should be swapped out?

How _should_ we deal with converters that accept types other than those in the registry?
  - should we call `convert` using the existing converter first to validate before overriding the converter? (See below after I finish testing this idea a bit more)
  
Should we be allowing the behavior #23417 allows, where units are set before data exists?
  - I can see why users might _want_ to do do this, but we have no lookup of which converter to use for a given unit, so it still feels fragile.

**Comment by mroeschke** [comment]
> The date converters have a default [unit] of None (should probably be allowed)

Sorry for a naive question, but what does unit as a `tzinfo` represent? And what would be the impact of `unit=None`  a "basic" unit like UTC if it was the unit of a naive datetimes?

> When handed strings in xlim, we say "Oh, you want a string categorical converter"

Is there a way for us (pandas) to override the xlim behavior and attempt to parse strings as a datetime, especially if the axis has datetime-like data? This would be an acceptable workaround IMO

**Comment by ksunden** [comment]
To your second question:

Not easily, especially not if you want it to only affect axes with Pandas Converters...

But we could do the following, which is kind of getting at that idea:

```diff
diff --git a/lib/matplotlib/axis.py b/lib/matplotlib/axis.py
index e836a329ba..6c19ad7097 100644
--- a/lib/matplotlib/axis.py
+++ b/lib/matplotlib/axis.py
@@ -1673,12 +1673,17 @@ class Axis(martist.Artist):
         ``axis.converter`` instance if necessary. Return *True*
         if *data* is registered for unit conversion.
         """
-        converter = munits.registry.get_converter(data)
-        if converter is None:
+        existing_converter = self.converter
+        try:
+            data = self.convert_units(data)
+        except Exception:
+            pass
+
+        if self.converter is None:
             return False
 
-        neednew = self.converter != converter
-        self.converter = converter
+        neednew = self.converter != existing_converter
+
         default = self.converter.default_units(data, self)
         if default is not None and self.units is None:
             self.set_units(default)
```

This diff fixes pandas, but breaks the testing vendored jpl_units (by passing bare floats to converters that expect labeled data... may be able to tweak it to fix that)

Not sure I _like_ this (especially the overly broad except, but converters can and do raise all kinds of errors...)

<details>
<summary> Summary of failing tests plus one relatively representative trace of most common failure </summary>

======================================================================== short test summary info =========================================================================
FAILED lib/matplotlib/tests/test_axes.py::test_formatter_ticker[pdf] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_axes.py::test_axhspan_epoch[png] - AttributeError: 'float' object has no attribute 'frame'
FAILED lib/matplotlib/tests/test_axes.py::test_formatter_ticker[svg] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_axes.py::test_formatter_ticker[png] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_axes.py::test_date_timezone_x_and_y[png] - matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 9.567):
FAILED lib/matplotlib/tests/test_axes.py::test_date_timezone_x[png] - matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 2.740):
FAILED lib/matplotlib/tests/test_axes.py::test_axhspan_epoch[pdf] - AttributeError: 'float' object has no attribute 'frame'
FAILED lib/matplotlib/tests/test_category.py::TestPlotTypes::test_update_plot[plot] - TypeError: 'value' must be an instance of str or bytes, not a float
FAILED lib/matplotlib/tests/test_axes.py::test_axvspan_epoch[pdf] - AttributeError: 'float' object has no attribute 'frame'
FAILED lib/matplotlib/tests/test_axes.py::test_axvspan_epoch[svg] - AttributeError: 'float' object has no attribute 'frame'
FAILED lib/matplotlib/tests/test_axes.py::test_fill_units[png] - AttributeError: 'float' object has no attribute 'frame'
FAILED lib/matplotlib/tests/test_category.py::test_overriding_units_in_plot[png] - TypeError: 'value' must be an instance of str or bytes, not a float
FAILED lib/matplotlib/tests/test_axes.py::test_axvspan_epoch[png] - AttributeError: 'float' object has no attribute 'frame'
FAILED lib/matplotlib/tests/test_axes.py::test_axhspan_epoch[svg] - AttributeError: 'float' object has no attribute 'frame'
FAILED lib/matplotlib/tests/test_patches.py::test_units_rectangle[png] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_axes.py::test_date_timezone_y[png] - matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 2.643):
FAILED lib/matplotlib/tests/test_polar.py::test_polar_units_2[png] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_polar.py::test_polar_units_1[pdf] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_polar.py::test_polar_units_2[svg] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_polar.py::test_polar_units_1[svg] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_polar.py::test_polar_units_1[png] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_units.py::test_jpl_barh_units[png] - AttributeError: 'float' object has no attribute 'frame'
FAILED lib/matplotlib/tests/test_units.py::test_jpl_bar_units[png] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_units.py::test_shared_axis_categorical - TypeError: 'value' must be an instance of str or bytes, not a float
FAILED lib/matplotlib/tests/test_units.py::test_unit_axis_label[plot] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_polar.py::test_polar_units_2[pdf] - AttributeError: 'float' object has no attribute 'type'
FAILED lib/matplotlib/tests/test_units.py::test_unit_axis_label[scatter] - AttributeError: 'float' object has no attribute 'type'
===================================================================== 28 failed, 39 passed in 8.16s ======================================================================


_____________________________________________________________________ test_unit_axis_label[scatter] ______________________________________________________________________
[gw3] linux -- Python 3.11.0 /home/kyle/.pyenv/versions/3.11.0/envs/edge/bin/python3.11

plot_meth_name = 'scatter'

    @pytest.mark.parametrize('plot_meth_name', ['scatter', 'plot'])
    def test_unit_axis_label(plot_meth_name):
        # Check that the correct Axis labels are set on plots with units
        import matplotlib.testing.jpl_units as units
        units.register()
    
        fig, ax = plt.subplots()
        ax.xaxis.set_units('m')
        ax.yaxis.set_units('sec')
        plot_method = getattr(ax, plot_meth_name)
>       plot_method(np.arange(3) * units.m, np.arange(3) * units.sec)

lib/matplotlib/tests/test_units.py:298: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lib/matplotlib/__init__.py:1442: in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
lib/matplotlib/axes/_axes.py:4579: in scatter
    x, y = self._process_unit_info([("x", x), ("y", y)], kwargs)
lib/matplotlib/axes/_base.py:2551: in _process_unit_info
    axis.update_units(data)
lib/matplotlib/axis.py:1687: in update_units
    default = self.converter.default_units(data, self)
lib/matplotlib/testing/jpl_units/UnitDblConverter.py:85: in default_units
    return UnitDblConverter.default_units(value[0], axis)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

value = 0.0, axis = <matplotlib.axis.XAxis object at 0x7f2d46c5dc90>

    @staticmethod
    def default_units(value, axis):
        # docstring inherited
        # Determine the default units based on the user preferences set for
        # default units when printing a UnitDbl.
        if cbook.is_scalar_or_string(value):
>           return UnitDblConverter.defaults[value.type()]
E           AttributeError: 'float' object has no attribute 'type'

lib/matplotlib/testing/jpl_units/UnitDblConverter.py:83: AttributeError


</details>

**Comment by ksunden** [comment]
To your first question:

Matplotlib units are not required to be strings, for all of the date converter, the thing that you are converting between is timezones, and `tzinfo` is the stdlib way of representing timezones, so that is the actual unit that the DateConverter (and friends, including Pandas' subclasses) The setting of units to the string `"UTC"` as I suggested early on in this issue is actually shorthand for setting to `datetime.timezone.utc` which is a tzinfo object (well, subclass thereof)

 `None` is the default tzinfo for `datetime`... it's effectively used to say "I don't want to do timezone conversions", and is sometimes normalized to be "UTC", or "machine local time" (though honestly not super standard)
 
 For matplotlib, a unit of `None` for date based converters means "use the timezone specified by `rcParams["timezone"]`", which defaults to UTC, but can be overridden by a user

The current implementation accepts None, and does not resolve it to the rcParam until trying to actually convert. The first diff above resolves that None when setting the default unit.

I think this is likely acceptable for Dates, but not sure a similar problem won't arise for other downstream libraries which provide converters, as None is a pretty popular choice for default units (it is in fact, what the base class for all unit converters returns, converters like the built-in DecimalConverter don't actually deal with units, per se, but use the units machinery to normalize decimal.Decimal to floats)

**Comment by tacaswell** [comment]
Pragmatic solution here may be to revert #23417 for 3.7.1 and then sort out the full fix on main?

**Comment by jklymak** [comment]
I think reverting #23417 makes the most sense - units=None clearly does not have the meaning of "converter not set", so any change that makes it have that meaning probably needs some sort of management.

**Comment by ksunden** [comment]
Closing as we reverted the change that caused this.

**Comment by mroeschke** [comment]
Thank you for the resolution!
